common/cross_sockets/XSocketServer.cpp

```diff
@@ -0,0 +1,134 @@
+#include "XSocketServer.h"
+
+#include "third-party/fmt/core.h"
+
+#include "common/cross_sockets/XSocket.h"
+
+#ifdef _WIN32
+#define NOMINMAX
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#include <WinSock2.h>
+#include <WS2tcpip.h>
+#endif
+
+XSocketServer::XSocketServer(std::function<bool()> shutdown_callback,
+                             int _tcp_port,
+                             int _buffer_size)
+    : want_exit_callback(std::move(shutdown_callback)) {
+  tcp_port = _tcp_port;
+  buffer_size = _buffer_size;
+  buffer = new char[_buffer_size];
+}
+
+XSocketServer::~XSocketServer() {
+  shutdown_server();
+}
+
+void XSocketServer::shutdown_server() {
+  // Close the listening and accepted socket socket
+  close_server_socket();
+  close_socket(accepted_socket);
+
+  // If the accept thread is still running (nothing ever connected)
+  // kill it and clean it up
+  if (accept_thread_running) {
```

Comment:
maybe this should go before the closes, so the thread doesn't try to `accept` on a socket that we just closed.